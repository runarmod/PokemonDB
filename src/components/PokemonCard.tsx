import { useQuery } from "@tanstack/react-query";
import { PokeAPI } from "pokeapi-types";
import { Requests } from "../api/Requests";
import Star from "../assets/star.png";
import FilledStar from "../assets/star_filled.png";
import "../styles/PokemonCard.css";
import { capitalizeFirstLetter, useAppContext } from "../utils";
import PokemonTypeLabel from "./PokemonTypeLabel";

const PokemonCard = () => {
    const { selectedPokemonId, favorites, updateFavorites } = useAppContext();

    function handleFavorite(): void {
        let updatedFavorites: number[] = [];
        if (favorites.includes(selectedPokemonId)) {
            updatedFavorites = favorites.filter(
                (favorite: number) => favorite !== selectedPokemonId
            );
        } else {
            updatedFavorites = [...favorites, selectedPokemonId];
        }
        updateFavorites(updatedFavorites);
    }

    function formatId(id: number): string {
        return `#${id.toString().padStart(4, "0")}`;
    }

    function removeDuplicateAbility(
        abilities: PokeAPI.PokemonAbility[]
    ): string[] {
        return [...new Set(abilities.map((a) => a.ability.name))];
    }

    function fetchData(): Promise<PokeAPI.Pokemon> | null {
        if (!selectedPokemonId) {
            return null;
        }
        return Requests.getPokemonById(selectedPokemonId);
    }

    const { data, error, isLoading } = useQuery({
        queryKey: ["pokemon", "id", selectedPokemonId],
        queryFn: fetchData,
    });

    if (isLoading) return <div>Loading...</div>;
    if (error) return <div>Error fetching Pokémon data</div>;
    if (data == undefined) {
        return (
            <article id="PokemonCardContainer">
                <p>
                    No Pokémon matches the filters. Please load more Pokémon or
                    change the filters.
                </p>
            </article>
        );
    }

    return (
        <article id="PokemonCardContainer">
            <figure id="ImageContainer">
                <button id="favoritesButton" onClick={handleFavorite}>
                    <img
                        id="StarIcon"
                        src={
                            favorites.includes(selectedPokemonId)
                                ? FilledStar
                                : Star
                        }
                        alt="Star image"
                    />
                </button>
                <img
                    id="PokemonImage"
                    src={data.sprites.front_default}
                    alt={`${capitalizeFirstLetter(data.name)} image`}
                />
            </figure>
            <div id="StatContainer">
                <section>
                    <h1>{capitalizeFirstLetter(data.name)}</h1>
                    <h2 style={{ opacity: 0.5 }}>
                        {formatId(selectedPokemonId)}
                    </h2>
                </section>
                <div id="StatGrid">
                    <section>
                        <h3>Height</h3>
                        <p>{data.height * 10} cm</p>
                    </section>
                    <section>
                        <h3>Weight</h3>
                        <p>{data.weight / 10} kg</p>
                    </section>
                    <section>
                        <h3>Types</h3>
                        <ul>
                            {data.types.map((pokemonType) => (
                                <li
                                    key={pokemonType.type.name}
                                    className="typeListEntry"
                                >
                                    <PokemonTypeLabel
                                        type={pokemonType.type.name}
                                    />
                                </li>
                            ))}
                        </ul>
                    </section>
                    <section>
                        <h3>Abilities</h3>
                        <ul>
                            {removeDuplicateAbility(data.abilities).map(
                                (name) => (
                                    <li key={name}>
                                        {capitalizeFirstLetter(name)}
                                    </li>
                                )
                            )}
                        </ul>
                    </section>
                </div>
            </div>
        </article>
    );
};

export default PokemonCard;
