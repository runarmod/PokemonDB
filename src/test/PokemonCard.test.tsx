import { useQuery } from "@tanstack/react-query";
import "@testing-library/jest-dom";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, Mock, vi } from "vitest";
import PokemonCard from "../components/PokemonCard";
import { useAppContext } from "../utils";
import mockPokemonData from "./mocks/mockPokemonData.json";

vi.mock(import("../utils"), async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...actual,
        useAppContext: vi.fn(),
    };
});

vi.mock(import("@tanstack/react-query"), async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...actual,
        useQuery: vi.fn(),
    };
});

describe("PokemonCard - General tests", () => {
    beforeEach(() => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            favorites: [],
            updateFavorites: vi.fn(),
        });
    });

    it("should display loading state", () => {
        (useQuery as Mock).mockReturnValue({
            isLoading: true,
            data: undefined,
            error: null,
        });
        render(<PokemonCard />);
        expect(screen.getByText("Loading...")).toBeInTheDocument();
    });

    it("should display error", () => {
        (useQuery as Mock).mockReturnValue({
            isLoading: false,
            data: undefined,
            error: new Error("This is an error"),
        });
        render(<PokemonCard />);
        expect(
            screen.getByText("Error fetching Pokémon data")
        ).toBeInTheDocument();
    });

    it("should display message if data is undefined", () => {
        (useQuery as Mock).mockReturnValue({
            isLoading: false,
            data: undefined,
            error: null,
        });
        render(<PokemonCard />);
        expect(
            screen.getByText(
                "No Pokémon matches the filters. Please load more Pokémon or change the filters."
            )
        ).toBeInTheDocument();
    });

    it("should display PokemonCard with stats", async () => {
        (useQuery as Mock).mockReturnValue({
            isLoading: false,
            data: mockPokemonData[0],
            error: null,
        });
        render(<PokemonCard />);

        // Check correct name
        expect(await screen.findByText("Bulbasaur")).toBeInTheDocument();

        // Check correct id formatting (tests formatID())
        expect(await screen.findByText("#0001")).toBeInTheDocument();

        // Check correct formatting of height and weight
        expect(await screen.findByText("70 cm")).toBeInTheDocument();
        expect(await screen.findByText("6.9 kg")).toBeInTheDocument();

        // Check types and abilities
        expect(await screen.findByText("grass")).toBeInTheDocument();
        expect(await screen.findByText("poison")).toBeInTheDocument();
        expect(await screen.findByText("Overgrow")).toBeInTheDocument();
        expect(await screen.findByText("Chlorophyll")).toBeInTheDocument();

        // Check Pokemon Image
        const pokemonImage = screen.getByAltText("Bulbasaur image");
        expect(pokemonImage).toBeInTheDocument();
        expect(pokemonImage).toHaveAttribute(
            "src",
            mockPokemonData[0].sprites.front_default
        );
    });

    it("should remove duplicate abilities (tests removeDuplicateAbility())", async () => {
        (useQuery as Mock).mockReturnValue({
            isLoading: false,
            data: {
                ...mockPokemonData[0],
                abilities: [
                    {
                        ability: { name: "overgrow" },
                    },
                    {
                        ability: { name: "overgrow" },
                    },
                ],
            },
            error: null,
        });
        render(<PokemonCard />);

        const abilityElements = await screen.findAllByText("Overgrow");
        expect(abilityElements).toHaveLength(1);
    });
});

describe("Snapshot test", () => {
    it("matches snapshot", () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            favorites: [0],
            updateFavorites: vi.fn(),
        });

        // Mocking of API call handled by handlers
        const { asFragment } = render(<PokemonCard />);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe("PokemonCard - Favorites test", () => {
    const mockUpdateFavorites = vi.fn();

    beforeEach(() => {
        mockUpdateFavorites.mockClear();
    });

    it("should display empty star when a Pokemon is not favorited", () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            favorites: [],
            updateFavorites: mockUpdateFavorites,
        });
        render(<PokemonCard />);

        const starIcon = screen.getByAltText("Star image");
        expect(starIcon).toHaveAttribute("src", "/src/assets/star.png");
    });

    it("should display empty star when other Pokemons are favorited", () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            favorites: [2],
            updateFavorites: mockUpdateFavorites,
        });
        render(<PokemonCard />);

        const starIcon = screen.getByAltText("Star image");
        expect(starIcon).toHaveAttribute("src", "/src/assets/star.png");
    });

    it("should display filled star when a Pokemon is favorited", () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            favorites: [1],
            updateFavorites: mockUpdateFavorites,
        });
        render(<PokemonCard />);

        const filledStarIcon = screen.getByAltText("Star image");
        expect(filledStarIcon).toHaveAttribute(
            "src",
            "/src/assets/star_filled.png"
        );
    });

    it("should display filled star when a Pokemon, amongst others, is favorited ", () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            favorites: [1, 2],
            updateFavorites: mockUpdateFavorites,
        });
        render(<PokemonCard />);

        const filledStarIcon = screen.getByAltText("Star image");
        expect(filledStarIcon).toHaveAttribute(
            "src",
            "/src/assets/star_filled.png"
        );
    });

    it("should add as favorite when clicking on empty favorite button", async () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            favorites: [],
            updateFavorites: mockUpdateFavorites,
        });
        render(<PokemonCard />);

        const favoriteButton = screen.getByRole("button", {
            name: "Star image",
        });
        userEvent.click(favoriteButton);

        await waitFor(() => {
            expect(mockUpdateFavorites).toHaveBeenCalledWith([1]);
        });
    });

    it("should be add to list of other favorites when clicking on empty favorite button", async () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            favorites: [2],
            updateFavorites: mockUpdateFavorites,
        });
        render(<PokemonCard />);

        const favoriteButton = screen.getByRole("button", {
            name: "Star image",
        });
        userEvent.click(favoriteButton);

        await waitFor(() => {
            expect(mockUpdateFavorites).toHaveBeenCalledWith([2, 1]);
        });
    });

    it("should be removed as favorite when clicking on filled favorite button", async () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            favorites: [1],
            updateFavorites: mockUpdateFavorites,
        });
        render(<PokemonCard />);

        const favoriteButton = screen.getByRole("button", {
            name: "Star image",
        });
        userEvent.click(favoriteButton);

        await waitFor(() => {
            expect(mockUpdateFavorites).toHaveBeenCalledWith([]);
        });
    });

    it("should remove only one Pokemon from favorite when clicking on filled favorite button", async () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            favorites: [1, 2],
            updateFavorites: mockUpdateFavorites,
        });
        render(<PokemonCard />);

        const favoriteButton = screen.getByRole("button", {
            name: "Star image",
        });
        userEvent.click(favoriteButton);

        await waitFor(() => {
            expect(mockUpdateFavorites).toHaveBeenCalledWith([2]);
        });
    });
});
