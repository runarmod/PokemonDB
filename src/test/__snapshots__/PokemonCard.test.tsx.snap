// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot test > matches snapshot 1`] = `
<DocumentFragment>
  <article
    id="PokemonCardContainer"
  >
    <figure
      id="ImageContainer"
    >
      <button
        id="favoritesButton"
      >
        <img
          alt="Star image"
          id="StarIcon"
          src="/src/assets/star.png"
        />
      </button>
      <img
        alt="Bulbasaur image"
        id="PokemonImage"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
      />
    </figure>
    <div
      id="StatContainer"
    >
      <section>
        <h1>
          Bulbasaur
        </h1>
        <h2
          style="opacity: 0.5;"
        >
          #0001
        </h2>
      </section>
      <div
        id="StatGrid"
      >
        <section>
          <h3>
            Height
          </h3>
          <p>
            70 cm
          </p>
        </section>
        <section>
          <h3>
            Weight
          </h3>
          <p>
            6.9 kg
          </p>
        </section>
        <section>
          <h3>
            Types
          </h3>
          <ul>
            <li
              class="typeListEntry"
            >
              <span
                class="pokemonTypeLabel"
                style="background-color: rgb(63, 161, 41);"
              >
                grass
              </span>
            </li>
            <li
              class="typeListEntry"
            >
              <span
                class="pokemonTypeLabel"
                style="background-color: rgb(145, 65, 203);"
              >
                poison
              </span>
            </li>
          </ul>
        </section>
        <section>
          <h3>
            Abilities
          </h3>
          <ul>
            <li>
              Overgrow
            </li>
          </ul>
        </section>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
